logic/jupyterkernel/stream.py

```diff
@@ -78,3 +89,167 @@ def close(self):
+        self._parent_header = value.copy()
+
+    @property
+    def port(self):
+        return self._port
+
+    @staticmethod
+    def bind(socket, connection, port):
+        if port <= 0:
+            return socket.bind_to_random_port(connection)
+        else:
+            socket.bind("%s:%s" % (connection, port))
+        return port
+
+    @staticmethod
+    def msg_id():
+        """ Return a new uuid for message id """
+        return str(uuid.uuid4())
+
+    def sign(self, msg_lst):
+        """
+        Sign a message with a secure signature.
+        """
+        h = self._auth.copy()
+        for m in msg_lst:
+            h.update(m)
+        return h.hexdigest().encode('ascii')
+
+    def new_header(self, msg_type):
+        """make a new header"""
+        return {
+            "date": datetime.datetime.now().isoformat(),
+            "msg_id": self.msg_id(),
+            "username": "kernel",
+            "session": self._engine_id,
+            "msg_type": msg_type,
+            "version": "5.0",
+        }
+
+    def send(self, msg_type, content=None, parent_header=None, metadata=None, identities=None):
+        header = self.new_header(msg_type)
+        if content is None:
+            content = dict()
+        if parent_header is None:
+            parent_header = self._parent_header
+        if metadata is None:
+            metadata = dict()
+
+        def jencode(msg):
+            return json.dumps(msg).encode('ascii')
+
+        msg_lst = [
+            jencode(header),
+            jencode(parent_header),
+            jencode(metadata),
+            jencode(content),
+        ]
+        signature = self.sign(msg_lst)
+        parts = [self.DELIM,
+                 signature,
+                 msg_lst[0],
+                 msg_lst[1],
+                 msg_lst[2],
+                 msg_lst[3]]
+        if identities:
+            parts = identities + parts
+        logging.debug('{0!s} send parts: {1!s}'.format(self.name, parts))
+        self.socket.send_multipart(parts)
+
+    def deserialize_wire_msg(self, wire_msg):
+        """split the routing prefix and message frames from a message on the wire"""
+        delim_idx = wire_msg.index(self.DELIM)
+        identities = wire_msg[:delim_idx]
+        m_signature = wire_msg[delim_idx + 1]
+        msg_frames = wire_msg[delim_idx + 2:]
+
+        def jdecode(msg):
+            return json.loads(msg.decode('ascii'))
+
+        m = {'header': jdecode(msg_frames[0]), 'parent_header': jdecode(msg_frames[1]),
+             'metadata': jdecode(msg_frames[2]), 'content': jdecode(msg_frames[3])}
+        check_sig = self.sign(msg_frames)
+        if check_sig != m_signature:
+            raise ValueError("Signatures do not match")
+
+        return identities, m
+
+
+class IOStdoutNetworkStream:
+    _output_channel = 'stdout'
+
+    def __init__(self, network_stream):
+        self._network_stream = network_stream
+
+    def write(self, s):
```

Comment:
This function needs to make sure that it does not call send from a thread that the socket does not belong to.

print() is not generally thread safe, it just works most of the time and does not crash, but there are no guarantees about doubled characters or lines, so it does not matter if our stdout replacement is thread safe.

The underlying socket is different, mixing up stream contents is a bad thing and sockets should only ever be read and written in a single thread.

Reply:
Very good point. I fixed that, see below.